### DISCUSSION TITLE
Testing joins with MockDatabase

### DISCUSSION THREAD

**Discussion by cemoktra** [SeaQL/sea-orm#443] [Q&A] [answered]
Hi,

is it possible to mock more than simple selects/inserts/updates? Basically i would like to test a query that uses `find_linked` and `find_also_related`

Do i just add models for each table that is part of the query with `append_query_results`? How can i see what is missing?

Is there a way add data once and not the mocks for multiple calls?

**Comment by cemoktra** [comment, marked as answer]
It seems i will need special mocking structs for the join, cause `try_get` looks for an `A_id` but in my struct i have an `id`

**Reply by billy1624** [reply]
Yes, you could use `maplit::btreemap!` to construct model attribute with `A_` prefix. This would be easier for you.

https://github.com/SeaQL/sea-orm/blob/d7e61b09c9c6b50be7c12149e00ac7b81063fff0/src/query/json.rs#L148-L152

**Reply by cemoktra** [reply]
Is it possible to somehow have a function that generates such a map given two `Entity` structs? Like iterating all columns adding `A_` to the first and `B_` to the second

**Reply by cemoktra** [reply]
Just as a proposal:

```
fn entity_to_prefixed_btreemap<T>(model: T::Model, prefix: &str) -> BTreeMap<String, sea_orm::Value>
    where
        T: EntityTrait,
{
    let mut mapped = BTreeMap::new();
    for column in T::Column::iter() {
        mapped.insert(format!("{}{}", prefix, column.as_str()), model.get(column));
    }
    mapped
}
```

and then:
```
fn mock_join<T1, T2>(model1: T1::Model, model2: T2::Model) -> BTreeMap<String, sea_orm::Value>
    where
        T1: EntityTrait,
        T2: EntityTrait,        
{
    let mut map = entity_to_prefixed_btreemap::<T1>(model1, "A_");
    map.extend(entity_to_prefixed_btreemap::<T2>(model2, "B_"));
    map
}
```

**Reply by billy1624** [reply]
Good suggestions!
I opened an tracking issue for this #447
